Apiserver watch from storage without PrevKV option

shadowofs opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Currently, the Kubernetes API server uses etcd’s `PrevKV` watch option to retrieve both the new and previous values of a key on each watch event. Under the hood, etcd performs a full range lookup to fetch the previous key-value pair, and all such operations contend on a single read–write lock protecting the internal treeIndex. As cluster scale grows, this coupling means that high write or watch load can starve other etcd requests, leading to increased watch-event latencies and overall pressure on etcd.

When the `watchCache` is enabled in the Apiserver, the only use of a watch event’s `PrevKV` is for the `Reflector` to remove deleted objects from the cache. However, removing deleted objects only requires the object’s name and namespace, which are already included in the etcd key of a deletion event. Therefore, we can safely disable `PrevKV` when the `watchCache` is enabled.

We validated this change on a 5,000‑node production cluster. Disabling `PrevKV` yielded ≈ 50 % reduction in MVCC range operations within etcd:
![image](https://github.com/user-attachments/assets/f61f5ed6-287c-4f09-b9ea-7c17c879f824)

We also ran benchmarks in a test environment. Disabling PrevKV showed ≈ 20 % increase in overall API throughput:
![image](https://github.com/user-attachments/assets/3f251dfe-5bdd-491d-9221-f336f881bbc7)

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
a new featureGate `WatchFromStorageWithoutPrevKV` has been added to enable etcd watches without `prev-kv`
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

shadowofs committed:
feat(apiserver): add featureGate WatchFromStorageWithoutPrevKV

shadowofs commented:
/test all

xigang commented:
@shadowofs If WithPrevKV() is disabled, will it affect the behavior of the client-side Reflector? For example, will the Update callback function in the EventHandler still receive the previous data?

serathius commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go:
```diff
@@ -426,7 +426,7 @@ func startCompactorOnce(c storagebackend.TransportConfig, interval time.Duration
 }
 
-func newETCD3Storage(c storagebackend.ConfigForResource, newFunc, newListFunc func() runtime.Object, resourcePrefix string) (storage.Interface, DestroyFunc, error) {
+func newETCD3Storage(c storagebackend.ConfigForResource, newFunc, newListFunc func() runtime.Object, newTombFunc func(string) ([]byte, error), resourcePrefix string) (storage.Interface, DestroyFunc, error) {
```
Not sure why we are creating tomb objects in etcd3 and not watch cache? Should watch cache be responsible for filling up prevKV?

serathius commented:
I agree with principle as this is something I proposed in https://github.com/kubernetes/kubernetes/issues/130939. 
Would like to opinion from cc @wojtek-t @jpbetz @deads2k

serathius commented:
Not sure about implementation, @shadowofs can you clarify the idea of tomb? Maybe I'm missing something, but when I was thinking about implementing this I didn't think we would need it. Definitely implementation could be simplified if we could remove it.

Jefftree commented:
/triage accepted

shadowofs commented:
> Not sure about implementation, @shadowofs can you clarify the idea of tomb? Maybe I'm missing something, but when I was thinking about implementing this I didn't think we would need it. Definitely implementation could be simplified if we could remove it.

I add the tomb because `Reflector` need a `runtime.Object` to purge cached object when a delete event is received. With the tomb, we don't need to change the logic of the `Reflector` to do the trick. Here is the code: 
```
# https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/reflector.go#L859
func handleAnyWatch(
	ctx context.Context,
	start time.Time,
	w watch.Interface,
	store ReflectorStore,
	expectedType reflect.Type,
	expectedGVK *schema.GroupVersionKind,
	name string,
	expectedTypeName string,
	setLastSyncResourceVersion func(string),
	exitOnWatchListBookmarkReceived bool,
	clock clock.Clock,
	errCh chan error,
) (bool, error) {
	...

loop:
	for {
		select {
		case <-ctx.Done():
			return watchListBookmarkReceived, errorStopRequested
		case err := <-errCh:
			return watchListBookmarkReceived, err
		case event, ok := <-w.ResultChan():
			...
			switch event.Type {
			case watch.Added:
				...
			case watch.Modified:
				...
			case watch.Deleted:
				err := store.Delete(event.Object)
				if err != nil {
					utilruntime.HandleErrorWithContext(ctx, err, "Unable to delete watch event object from store", "reflector", name, "object", event.Object)
				}
```

shadowofs commented:
> @shadowofs If `WithPrevKV()` is disabled, will it affect the behavior of the client-side `Reflector`? For example, will the Update callback function in the `EventHandler` still receive the previous data?

No, that's the beauty of it. Because Client side `Reflector` watches from `watchCache` instead of etcd when `watchCache` is enabled.

xigang commented:
> > @shadowofs If `WithPrevKV()` is disabled, will it affect the behavior of the client-side `Reflector`? For example, will the Update callback function in the `EventHandler` still receive the previous data?
> 
> No, that's the beauty of it. Because Client side `Reflector` watches from `watchCache` instead of etcd when `watchCache` is enabled.

Got it, thanks a lot.

serathius commented:
>  So by adding the tomb, we don't need to change the Store interface.

I don't think it's that important.

shadowofs commented:
> Let's first get agreement about #130939
I aggreed and that's what I did in this PR. I can give you more cases why this is important. `WatchPrevKV` can cause etcd pending events rising due to treeIndex lock contension. 
![image](https://github.com/user-attachments/assets/d617c527-4111-4039-bb51-b034c971ae85)
We have ran into cases when some large range took over 1m time, all other requests including watch events delivery starved.
![image](https://github.com/user-attachments/assets/56d734a9-4a14-4ce7-9bd5-a2c9e8672f9f)

So no doubt getting rid of WatchCache is a good idea.

> I don't think it's that important. I think we should prioritize separation of concerns. I'm not convinced that etcd3 should be aware of tombFunc, nor I think piping it through all layers is good.
Only etcd3 can touch etcd events and the deleted key. I think exposing etcd key to the Cacher or the Reflector would be worse than adding a tombFunc in etcd3.

shadowofs commented:
> Let's first get agreement about #130939

I aggreed and that's what I did in this PR. I can give you more cases why this is important. `WatchPrevKV` can cause etcd pending events rising due to treeIndex lock contension.
![image](https://private-user-images.githubusercontent.com/681144/445980959-d617c527-4111-4039-bb51-b034c971ae85.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDc4MTczMzQsIm5iZiI6MTc0NzgxNzAzNCwicGF0aCI6Ii82ODExNDQvNDQ1OTgwOTU5LWQ2MTdjNTI3LTQxMTEtNDAzOS1iYjUxLWIwMzRjOTcxYWU4NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUyMVQwODQzNTRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xYTcyMzgyMmZhYjU4MGYyYzFlMDMwMDMzODMxYzljNTFkMjg5MTcyMDg4ZmNhMDE4M2MxMGRjZjZhNjZmZThiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.u-CGdjkXaEHBqGuHgJc1xN2mF7h1AldzdIK93Tbe46k)
We have ran into cases when some large range took over 1m time, all other requests including watch events delivery starved.
![image](https://private-user-images.githubusercontent.com/681144/445983956-56d734a9-4a14-4ce7-9bd5-a2c9e8672f9f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDc4MTczMzQsIm5iZiI6MTc0NzgxNzAzNCwicGF0aCI6Ii82ODExNDQvNDQ1OTgzOTU2LTU2ZDczNGE5LTRhMTQtNGNlNy05YmQ1LWEyYzllODY3MmY5Zi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUyMVQwODQzNTRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hZDlmZTkyMGRhNGYzNjY2NGQ4MmUxNWJkMjJkZGFlOWY3NDQxZjIyYmM0YTkzMTg2MDJkNjU5MzZmMGZkOGZiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.2iJ0u6C9mzP0xYyGyf6ozZe5qpY6MKIviDGlaUgUwbU)

So no doubt getting rid of WatchCache is a good idea.

> I don't think it's that important. I think we should prioritize separation of concerns. I'm not convinced that etcd3 should be aware of tombFunc, nor I think piping it through all layers is good.

Only etcd3 can touch etcd events and the deleted key. I think exposing etcd key to the Cacher or the Reflector would be worse than adding a tombFunc in etcd3.

serathius commented:
> So no doubt getting rid of PrevKV is a good idea.

I have no doubt of the benefits, however this is not a trivial change and we need to have consensus with the SIG-Apimachinery leads and storage approvers.

shadowofs commented:
> > So no doubt getting rid of PrevKV is a good idea.
> 
> I have no doubt of the benefits, however this is not a trivial change and we need to have consensus with the SIG-Apimachinery leads and storage approvers.

Got it. Thanks for the reply.

serathius commented:
Got +1 from @wojtek-t https://github.com/kubernetes/kubernetes/issues/130939#issuecomment-2897240499

shadowofs commented:
> Not sure about implementation, @shadowofs can you clarify the idea of tomb? Maybe I'm missing something, but when I was thinking about implementing this I didn't think we would need it. Definitely implementation could be simplified if we could remove it.

@serathius I forget one thing about this. The keyFunc and etcd key prefix are configurable. To be specified, for namespace-scoped resource the default key format is /<prefix>/<resource>/<namespace>/<name>. While for cluster-scoped resource the default key format is /<prefix>/<resource>/<name>. So the tombFunc needs to be all the way passed down from Store, where the keyFunc is defined.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/event.go:
```diff
@@ -70,6 +70,12 @@ func parseEvent(e *clientv3.Event) (*event, error) {
 	}
 	if e.PrevKv != nil {
 		ret.prevValue = e.PrevKv.Value
+	} else if e.Type == clientv3.EventTypeDelete && newTombFunc != nil {
```
Actually, we probably need to pass the `reversedKeyFunc(string) (string, string, error)` (to split the name into namespace, name) - but I think it's still better to pass that, then the concept of tombstone function.

xigang commented:
Keep most of the changes scoped to the store layer?
That way, we limit the impact of the change and avoid modifying higher-level logic like the watch cache. For example:

Modify the Watch method in the store implementation:

```
// Watch implements storage.Interface.Watch.
func (s *store) Watch(ctx context.Context, key string, opts storage.ListOptions) (watch.Interface, error) {
    preparedKey, err := s.prepareKey(key)
    if err != nil {
        return nil, err
    }
    rev, err := s.versioner.ParseResourceVersion(opts.ResourceVersion)
    if err != nil {
        return nil, err
    }

    watchOpts := kubernetes.WatchOptions{
        Revision: int64(rev),
        PrevKV: !utilfeature.DefaultFeatureGate.Enabled(features.WatchFromStorageWithoutPrevKV),
    }

    return s.watcher.WatchWithOptions(s.watchContext(ctx), preparedKey, watchOpts, opts)
}
```

And in the watcher implementation:
```
func (w *watcher) WatchWithOptions(ctx context.Context, key string, watchOpts kubernetes.WatchOptions, opts storage.ListOptions) (watch.Interface, error) {
    // Use watchOpts.PrevKV to decide whether we need the previous value
    watchOptions := []clientv3.OpOption{
        clientv3.WithRev(watchOpts.Revision),
    }
    if watchOpts.PrevKV {
        watchOptions = append(watchOptions, clientv3.WithPrevKV())
    }

    // Proceed with the rest of the watch logic
    ...
}

```

shadowofs commented:
@xigang I don't think it's a good idea to store layer interface. `PrevKV` is not a generic feature in store layer. See the `Watch` interface: 
```
// Interface can be implemented by anything that knows how to watch and report changes.
type Interface interface {
	// Stop tells the producer that the consumer is done watching, so the
	// producer should stop sending events and close the result channel. The
	// consumer should keep watching for events until the result channel is
	// closed.
	//
	// Because some implementations may create channels when constructed, Stop
	// must always be called, even if the consumer has not yet called
	// ResultChan().
	//
	// Only the consumer should call Stop(), not the producer. If the producer
	// errors and needs to stop the watch prematurely, it should instead send
	// an error event and close the result channel.
	Stop()

	// ResultChan returns a channel which will receive events from the event
	// producer. If an error occurs or Stop() is called, the producer must
	// close this channel and release any resources used by the watch.
	// Closing the result channel tells the consumer that no more events will be
	// sent.
	ResultChan() <-chan Event
}

// Event represents a single event to a watched resource.
// +k8s:deepcopy-gen=true
type Event struct {
	Type EventType

	// Object is:
	//  * If Type is Added or Modified: the new state of the object.
	//  * If Type is Deleted: the state of the object immediately before deletion.
	//  * If Type is Bookmark: the object (instance of a type being watched) where
	//    only ResourceVersion field is set. On successful restart of watch from a
	//    bookmark resourceVersion, client is guaranteed to not get repeat event
	//    nor miss any events.
	//  * If Type is Error: *api.Status is recommended; other types may make sense
	//    depending on context.
	Object runtime.Object
} 
```
The Event struct has only one Object. The notion `PrevKV` is sealed inside store layer. The interface end-user should never understand what is a `PrevKV` unless we want to change the watch protocol.

This should be an etcd3 feature. But the problem is etcd3 doesn't know how to deal with the etcd key. Because the etcd key format is defined by two things: 
- `Prefix` in the `StorageConfig` which by default is `/registry`
- `Store.KeyFunc` which by default can be either `/<resource>/<namespace>/<name>` or `/<resource>/<name>` depending on the scope of the resource.

Both the two things are configurable. As a result, only `Store` has the knowledge how to decode the etcd key, which makes me introduce this "tedious" `tombFunc`, and passes it all the way down from where the `KeyFunc` is defined.

xigang commented:
Since the `store` layer already owns both `keyFunc` and `newFunc`, and knows the key prefix and resource scope, we can move the reverse-parsing logic directly into `prepareObjs()` by constructing a `reverseKeyFunc` internally. This keeps the key semantics encapsulated within the store, avoids polluting interfaces with tombFunc, and reduces the amount of change needed across the stack.


Let’s see the feedback from others.

serathius commented:
/ok-to-test

shadowofs commented:
/test pull-kubernetes-linter-hints

shadowofs commented:
/retest

dims commented on staging/src/k8s.io/apiserver/pkg/features/kube_features.go:
```diff
@@ -407,6 +411,10 @@ var defaultVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate
 	},
 
+	WatchFromStorageWithoutPrevKV: {
+		{Version: version.MustParse("1.33"), Default: false, PreRelease: featuregate.Alpha},
```
`1.33` already shipped. please use `1.34`

dims commented on test/compatibility_lifecycle/reference/versioned_feature_list.yaml:
```diff
@@ -1681,6 +1681,12 @@
+  - default: false
+    lockToDefault: false
+    preRelease: Alpha
+    version: "1.33"
```
1.34

dims commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -1597,6 +1597,30 @@ func (e *Store) CompleteWithOptions(options *generic.StoreOptions) error {
+		tokens := strings.Split(key, prefix)
+		if len(tokens) != 2 {
+			err = fmt.Errorf("invalid key %q, requiring resource prefix %q", key, prefix)
+			return
```
can you please use explicit return with all the 3 values in all the empty `return`(s)? makes it easier to read

shadowofs commented:
/retest

shadowofs commented:
/retest

shadowofs commented:
/retest-required

dims commented:
/retest

shadowofs commented:
/test pull-kubernetes-node-e2e-containerd

serathius commented on pkg/features/kube_features.go:
```diff
@@ -1338,6 +1338,10 @@ var defaultVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate
 	},
 
+	genericfeatures.WatchFromStorageWithoutPrevKV: {
+		{Version: version.MustParse("1.34"), Default: false, PreRelease: featuregate.Alpha},
```
This feature should launch as Beta default, as we are not implementing a KEP, rather are using feature gate as an escape hatch. Of course this means we will need to ensure Beta quality out of the gate.

serathius commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go:
```diff
@@ -464,7 +464,10 @@ func newETCD3Storage(c storagebackend.ConfigForResource, newFunc, newListFunc fu
 	versioner := storage.APIObjectVersioner{}
 	decoder := etcd3.NewDefaultDecoder(c.Codec, versioner)
-	store := etcd3.New(client, c.Codec, newFunc, newListFunc, c.Prefix, resourcePrefix, c.GroupResource, transformer, c.LeaseManagerConfig, decoder, versioner)
+	if c.WatchPrevKv {
```
I don't like the fact that we set WatchPrevKv per storage. This changes the results returned in watch without user knowledge and makes watchcache and etcd behavior inconsistent. It only passes tests because on implicit behavior that Watch cache doesn't pass any requests to etcd. If we ever changed that, there would be very bad consequences.

In my opinion the Storage interface was designed to ensure 1to1 parity of behavior of etcd3 and cache. I would prefer PrevKV to be a parameter in Options passed to Watch, so this is requested by client. This would make so that both behave the same if PrevKV is passed, and result changes when PrevKV is set to false. 

cc @wojtek-t  for opinion.

shadowofs replied:
> I don't like the fact that we set WatchPrevKv per storage. This changes the results returned in watch without user knowledge and makes watchcache and etcd behavior inconsistent.

WatchPrevKv is set per storage due to WatchCache is set per storage. So it happens when two Storage with different WatchCache settings should be set with different WatchPrevKv settings. 

And I don't think watchcache should be consistent with etcd behavior. We have never promised that. Otherwise, Sotrage watch event should add a prevObject value. 

> It only passes tests because on implicit behavior that Watch cache doesn't pass any requests to etcd.

It passes the tests because we haven't changed the behavior of `WatchCache` as a Storage implementation. In other words, `Storage` clients don't care about `WatchPrevKv`. The `WatchPrevKv` logic should be encapsulated within the etcd `Storage` implementation.

> In my opinion the Storage interface was designed to ensure 1to1 parity of behavior of etcd3 and cache.

I am sorry I can't see the benifit of doing so right now. Maybe you can be more specific why we should do so. By the  way, k3s uses `SqlLite` as a `Storage`.

serathius replied:
> WatchPrevKv is set per storage due to WatchCache is set per storage. So it happens when two Storage with different WatchCache settings should be set with different WatchPrevKv settings.

If the `WatchPrevKv` was passed by the caller, it would be configured on WatchCache and not on Storage. 

> And I don't think watchcache should be consistent with etcd behavior. We have never promised that. Otherwise, Sotrage watch event should add a prevObject value.

Who are "we" in this statement? etcd3 and watchcache implement the same interface because they are meant to provide exactly the same behavior. By behavior I mean for the same input it should return the same output with the same consistency. This is why we have generic storage tests that run on both. While not every behavior is user facing, we do that to make maintenance easier. 

My prefer having `WatchPrevKv` in Watch request, because it makes the new behavior explicit to client instead of depending on implicit configuration. But would like to get opinion from @wojtek-t @deads2k or @jpbetz 

> It passes the tests because we haven't changed the behavior of WatchCache as a Storage implementation. In other words, Storage clients don't care about WatchPrevKv.

They care, the watchcache is a client of etcd storage.

> I am sorry I can't see the benifit of doing so right now. Maybe you can be more specific why we should do so. By the way, k3s uses SqlLite as a Storage.

This is K8s not k3s.

shadowofs replied:
> If the WatchPrevKv was passed by the caller, it would be configured on WatchCache and not on Storage.

So you mean watchcache should not start caching until a client called `Watch`? What if the client just called List.

serathius replied:
> So you mean watchcache should not start caching until a client called Watch? What if the client just called List.

No, I mean that requests made by ListWatcher, which is part of watchcache, should pass `WatchPrevKV=false`. While direct watches made by rest API or other clients should pass `WatchPrevKV=true`, because they are expecting PrevKV.

shadowofs commented:
/retest

shadowofs commented:
> ```
> === RUN   TestAllocateReservedDynamicBlockExhausted
>     storage_test.go:59: unexpected error creating etcd: watch prevkv disabled without reverseKeyFunc
> ```

I'll fix this later

shadowofs commented:
/retest

xigang commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features

serathius commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -1597,6 +1597,30 @@ func (e *Store) CompleteWithOptions(options *generic.StoreOptions) error {
+	// Examples:
+	//		key:    /registry/pods/namespace/name
+	//		prefix: /pods
+	reverseKeyFunc := func(key string) (name string, namespace string, err error) {
```
This can be normal function, not an anonymous one. This would allow us to unit test with `NoNamespaceKeyFunc` and `NamespaceKeyRootFunc` like `reverseKeyFunc(NoNamespaceKeyFunc(x)) == x`

serathius commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -1427,12 +1427,12 @@ func (e *Store) Watch(ctx context.Context, options *metainternalversion.ListOpti
 // WatchPredicate starts a watch for the items that matches.
-func (e *Store) WatchPredicate(ctx context.Context, p storage.SelectionPredicate, resourceVersion string, sendInitialEvents *bool) (watch.Interface, error) {
-	storageOpts := storage.ListOptions{ResourceVersion: resourceVersion, Predicate: p, Recursive: true, SendInitialEvents: sendInitialEvents}
+func (e *Store) WatchPredicate(ctx context.Context, p storage.SelectionPredicate, resourceVersion string, sendInitialEvents *bool, withoutPrevKV bool) (watch.Interface, error) {
```
Not sure why this is an argument, could we just check feature flag?

serathius commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:
```diff
@@ -419,6 +421,11 @@ func (f *StorageFactoryRestOptionsFactory) GetRESTOptions(resource schema.GroupR
 			klog.V(3).InfoS("Using watch cache", "resource", resource)
 			ret.Decorator = genericregistry.StorageWithCacher()
+
+			if utilfeature.DefaultFeatureGate.Enabled(features.WatchFromStorageWithoutPrevKV) {
```
This is not needed, apiserver logs command line flags including feature flags that were set.

serathius replied:
Right, so it will generate a lot of logs that all will say the same value. Whether `WatchFromStorageWithoutPrevKV` really works is based on fact whether watch cache is enabled and feature flag is set.

serathius commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -120,6 +120,10 @@ type Config struct {
+
+	// PrevKV is the previous generation of the object in etcd.
+	// When WatchWithoutPrevKV is enabled, the prevKV field in etcd watch events will be excluded.
+	WatchWithoutPrevKV bool
```
This should not be passed via config.

serathius commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go:
```diff
@@ -69,9 +70,11 @@ func TestOnlySetFatalOnDecodeError(b bool) {
+	codec   runtime.Codec
+	newFunc func() runtime.Object
+	// reverseKeyFunc decodes storage key to get object name and namespace
+	reverseKeyFunc      func(key string) (string, string, error)
```
Declare type 
```
type ReverseKeyFunc func(key string) (namespace string, name string, error)
```

serathius commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go:
```diff
@@ -706,16 +712,32 @@ func (wc *watchChan) prepareObjs(e *event) (curObj runtime.Object, oldObj runtim
-		oldObj, err = decodeObj(wc.watcher.codec, wc.watcher.versioner, data, e.rev)
-		if err != nil {
-			return nil, nil, wc.watcher.transformIfCorruptObjectError(e, err)
+	if e.isDeleted || !wc.acceptAll() {
```
If predicate is not empty, we need to lookup the prevKV to be able to determine if the object from prevKV fulfilled the predicate to decide whether we should inform watcher or drop.

shadowofs commented:
@wojtek-t is currently on paternity leave, so kindly asking if @aojea, @cjcullen, @dims, or @serathius could help review this PR. Thanks a lot!

serathius commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -306,6 +308,26 @@ func NoNamespaceKeyFunc(ctx context.Context, prefix string, name string) (string
+// context does not contain a namespace, it errors.
+func NamespaceReverseKeyFunc(key string) (name string, namespace string, err error) {
+	tokens := strings.Split(key, "/")
+	if len(tokens) < 2 {
```
What happens if `len(tokens) > 2` we should return error instead of returning invalid response.

serathius commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -1435,7 +1457,13 @@ func (e *Store) Watch(ctx context.Context, options *metainternalversion.ListOpti
+		Predicate:          p,
+		Recursive:          true,
+		SendInitialEvents:  sendInitialEvents,
+		WatchWithoutPrevKV: utilfeature.DefaultFeatureGate.Enabled(features.WatchFromStorageWithoutPrevKV),
```
This is incorrect place to set `WatchWithoutPrevKV`. It should only be used by cacher.

serathius commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/event.go:
```diff
@@ -55,11 +55,10 @@ func parseKV(kv *mvccpb.KeyValue) *event {
-func parseEvent(e *clientv3.Event) (*event, error) {
-	if !e.IsCreate() && e.PrevKv == nil {
+func parseEvent(e *clientv3.Event, checkPrevKV bool) (*event, error) {
+	if checkPrevKV && !e.IsCreate() && e.PrevKv == nil {
```
If we don't check `PrevKV` we should also not pass it on line 70.

serathius commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:
```diff
@@ -141,7 +141,7 @@ func (a *abortOnFirstError) Aggregate(key string, err error) bool {
 
 // New returns an etcd3 implementation of storage.Interface.
-func New(c *kubernetes.Client, codec runtime.Codec, newFunc, newListFunc func() runtime.Object, prefix, resourcePrefix string, groupResource schema.GroupResource, transformer value.Transformer, leaseManagerConfig LeaseManagerConfig, decoder Decoder, versioner storage.Versioner) *store {
+func New(c *kubernetes.Client, codec runtime.Codec, newFunc, newListFunc func() runtime.Object, reverseKeyFunc func(string) (string, string, error), prefix, resourcePrefix string, groupResource schema.GroupResource, transformer value.Transformer, leaseManagerConfig LeaseManagerConfig, decoder Decoder, versioner storage.Versioner) *store {
```
Please add tests for etcd3/store for `WatchWithoutPrevKV`

serathius commented:
> This failure is caused by timeout. I wonder is it possible to extend the timeout limit.

No, cacher package has been problematic with timeouts as adding new tests can easily lead to hitting the 3 minute limit. There were some attempts to split the package, but because the code is very intertwined it didn't succeed. SIG-testing was against increasing the 3 minute limit.

I would recommend ensuring that you ensure that you avoid unnecessary sleeps and waits in tests.

shadowofs commented:
@serathius updated. PTAL

shadowofs commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_test.go:
```diff
@@ -393,30 +393,42 @@ func TestStats(t *testing.T) {
 }
 
 func TestWatchFromZero(t *testing.T) {
+	// TODO(#131862): Test both WatchWithoutPrevKV=true and false
```
@serathius Moving these tests into TestWatch may end up in time out. So I added some TODO here. After we split this package, we can add more tests for WatchWithoutPrevKV.

serathius commented:
Will take a look, however I would prefer we merge safeguards into key patterns first. For 1.34 we had some issues with ensuring proper usage of "/" in key patterns. As followup I'm adding validation that ensures that keys passed to storage follow correct pattern. However it might take some time as there are a lot of tests that need to be fixed. For example https://github.com/kubernetes/kubernetes/pull/133871. 

This should make your change much safer. Hoping to get it into 1.35.

shadowofs commented:
Hi @serathius,
I hope you’re doing well. Just wanted to kindly check in on this PR.

Earlier you mentioned that safeguards for key patterns and the related validation work. Since it has been a few months, I wanted to ask whether there has been any progress on that work, and whether there’s anything I can help with to unblock or move this PR forward.

Please let me know if I can assist in any way. Thanks again for your time and review.

shadowofs commented:
Hi @serathius,

Thanks for the clarification and the detailed context.

I agree that the tests should cover both cases with the feature gate enabled and disabled. That makes sense to ensure the behavior is fully validated.

Regarding the proposal to split the package to avoid test timeouts, I think this is worth discussing, but it may be better handled separately. I suggest opening a dedicated issue to explore the options and design for the package restructuring, so we can move this PR forward without blocking on that discussion.

Let me know if you’d like me to help initiate that issue.

shadowofs commented:
Hi @serathius,

Thanks again for the earlier reviews. I know you were quite busy with the v1.35 release recently — congrats on the release 🎉

Since v1.35 has now been cut, I was wondering if there’s a good opportunity to move this PR forward. I’m happy to address any remaining feedback or make adjustments if needed.

cc @liggitt @wojtek-t for opinion.

Thanks!

serathius commented:
See https://github.com/kubernetes/kubernetes/issues/135960 for a proposal to split the cacher package to reduce test duration and unblock this PR.

shadowofs commented:
@liggitt @serathius PTAL.

serathius commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_test.go:
```diff
@@ -399,107 +399,185 @@ func TestKeySchema(t *testing.T) {
 }
 
 func TestWatch(t *testing.T) {
```
This file should be kept somewhat similar to https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher_test.go

Can you split this change it test structure to separate PR that updates both cacher and etcd3?

shadowofs replied:
Done. I reverted the structural change in cacher_test.go and kept the watch tests as separate `TestWatch* `functions, matching the layout in [etcd3/watcher_test.go](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher_test.go) more closely.

  I only kept a small targeted test for `WatchFromStorageWithoutPrevKV` to cover the new feature gate behavior without reorganizing the whole watch test suite.

serathius replied:
That's not what I asked for. The additional test coverage is important. We want to make sure all features of watch are covered for this feature flag.

shadowofs replied:
Thanks for clarifying. I misunderstood the previous comment as a request to remove the broader watcher coverage from this PR entirely.

I split the structural test refactor into a separate PR: https://github.com/kubernetes/kubernetes/pull/138638. Once that lands, I’ll rebase this PR and restore the feature-gate coverage across the full watch test suite on top of the shared structure, so the WatchFromStorageWithoutPrevKV path is covered for all watch behaviors without mixing the structural refactor into this PR.

serathius commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store_test.go:
```diff
@@ -2435,7 +2435,7 @@ func newTestGenericStoreRegistry(t *testing.T, scheme *runtime.Scheme, hasCacheE
 
 	sc.Codec = apitesting.TestStorageCodec(codecs, examplev1.SchemeGroupVersion)
-	s, dFunc, err := factory.Create(*sc.ForResource(schema.GroupResource{Resource: "pods"}), newFunc, newListFunc, "/pods")
+	s, dFunc, err := factory.Create(*sc.ForResource(schema.GroupResource{Resource: "pods"}), newFunc, newListFunc, NamespaceReverseKeyFunc(podPrefix), podPrefix)
```
```suggestion
	s, dFunc, err := factory.Create(*sc.ForResource(schema.GroupResource{Resource: "pods"}), newFunc, newListFunc, NamespaceReverseKeyFunc(podPrefix), "/pods")
```

serathius commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -511,10 +511,14 @@ func (c *Cacher) Watch(ctx context.Context, key string, opts storage.ListOptions
 	if err != nil {
 		return nil, err
 	}
+
```
```suggestion
```

serathius commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go:
```diff
@@ -74,9 +71,11 @@ func TestOnlySetFatalOnDecodeError(b bool) {
+	client  *clientv3.Client
+	codec   runtime.Codec
+	newFunc func() runtime.Object
+	// reverseKeyFunc decodes storage key to get object name and namespace
```
```suggestion
```

shadowofs commented:
@serathius all comments resolved. PTAL.

serathius commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -1620,6 +1661,14 @@ func (e *Store) CompleteWithOptions(options *generic.StoreOptions) error {
+	// Reverse the key generated by keyFunc into object name and namespace.
+	var reverseKeyFunc storage.ReverseKeyFunc
+	if isNamespaced {
+		reverseKeyFunc = NamespaceReverseKeyFunc(storageKeyPrefix)
```
KeyFunc and ReverseKeyFunc should be decided in symmetrical pairs. Can we combine this with section that sets `KeyFunc` and `KeyRootFunc` on 1630.

serathius commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store_test.go:
```diff
@@ -2490,10 +2490,11 @@ func newTestGenericStoreRegistry(t *testing.T, scheme *runtime.Scheme, hasCacheE
 			return podPrefix
 		},
 		KeyFunc: func(ctx context.Context, id string) (string, error) {
```
ditto. Can you move the `KeyFunc` creation next to line 2438 where we decide on `ReverseKeyFunc` by passing NamespaceReverseKeyFunc(podPrefix).

```
  keyFunc := func (ctx context.Context, id string) (string, error) {
if _, ok := genericapirequest.NamespaceFrom(ctx); !ok {
			if namespace, ok := genericapirequest.NamespaceFrom(ctx); !ok {
				return "", fmt.Errorf("namespace is required")
			} else {
				return path.Join(podPrefix, namespace, id), nil
			}
			return path.Join(podPrefix, id), nil
		}
  reverseKeyFunc := NamespaceReverseKeyFunc(podPrefix)
```

serathius commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_test.go:
```diff
@@ -399,92 +399,97 @@ func TestKeySchema(t *testing.T) {
-		t.Cleanup(terminate)
-		storagetesting.RunWatchListMatchSingle(context.TODO(), t, store)
-	})
+	for _, watchWithoutPrevKV := range []bool{true, false} {
```
Thanks for improving this, Now this is super easy to review with `?w=1` argument. Like https://github.com/kubernetes/kubernetes/pull/131862/changes?w=1

shadowofs replied:
Thanks! Glad it makes the review easier.

serathius commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_testing_utils_test.go:
```diff
@@ -56,6 +57,21 @@ func init() {
 
 func newPod() runtime.Object     { return &example.Pod{} }
 func newPodList() runtime.Object { return &example.PodList{} }
+func reverseKeyFunc(prefix string) storage.ReverseKeyFunc {
```
Can you move it next to matching KeyFunc.

serathius commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_testing_utils_test.go:
```diff
@@ -56,6 +57,21 @@ func init() {
 func newPod() runtime.Object     { return &example.Pod{} }
 func newPodList() runtime.Object { return &example.PodList{} }
+func reverseKeyFunc(prefix string) storage.ReverseKeyFunc {
+	return func(key string) (name string, namespace string, err error) {
```
For each KeyFunc and ReverseKeyFunc pair we should have a test.

serathius commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/event_test.go:
```diff
@@ -30,6 +30,7 @@ func TestParseEvent(t *testing.T) {
 		name          string
 		etcdEvent     *clientv3.Event
 		expectedEvent *event
+		checkPrevKV   bool
```
```suggestion
		expectPrevKV   bool
```

serathius commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go:
```diff
@@ -77,6 +77,16 @@ func newPodList() runtime.Object {
 	return &example.PodList{}
 }
 
+func reverseKeyFunc(key string) (name string, namespace string, err error) {
```
Can you move it next to matching KeyFunc?

serathius commented on staging/src/k8s.io/apiserver/pkg/storage/interfaces.go:
```diff
@@ -330,6 +330,10 @@ type ListOptions struct {
+	// PrevKV is the previous generation of the object in etcd.
+	// When WatchWithoutPrevKV is enabled, the prevKV field in etcd watch events will be excluded.
+	// The option is ignored for non-watch requests.
+	WatchWithoutPrevKV bool
```
How cacher should handle this option? Think just rejecting should be good enough.

shadowofs replied:
Done. Added rejection in cacher.Watch() — cacher serves from the watch cache and doesn't interact with etcd directly, so WatchWithoutPrevKV (which strips prevKV at the etcd layer) is not applicable here. Just returning an  error is the right call.                                                                                                                                                                                                         
                                                                                                                                                                                                                                   
```
if opts.WatchWithoutPrevKV {           
    return nil, fmt.Errorf("watchWithoutPrevKV is not supported by the cacher")                                                                                                                                                  
} 
```

shadowofs commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -1595,13 +1637,15 @@ func (e *Store) CompleteWithOptions(options *generic.StoreOptions) error {
 			e.KeyFunc = func(ctx context.Context, name string) (string, error) {
 				return NamespaceKeyFunc(ctx, prefix, name)
 			}
+			reverseKeyFunc = NamespaceReverseKeyFunc(storageKeyPrefix)
```
we shouldn't set reverseKeyFunc when keyFunc is customized

serathius commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
Looks like we have pairs of (keyFunc, reverseKeyFunc) for both (storage, cacher). Not sure which goes where and they handle namespacing differently (separate function vs isNamespaced passed via closure)

Could we cleanup layering here?

shadowofs replied:
I cleaned up the layering by making the decorator parameters explicit: `cacheKeyFunc` is only used by the cacher/watch cache to compute resource keys from objects, while `storageReverseKeyFunc` is passed to raw storage for recovering object identity from prepared storage keys.
 
I also moved the default namespaced/non-namespaced key function setup into `defaultStoreKeyFuncs`, so the default `KeyRootFunc`, `KeyFunc`, and reverse key func are generated together instead of being split across the storage/cacher setup path.

serathius commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -1586,39 +1669,14 @@ func (e *Store) CompleteWithOptions(options *generic.StoreOptions) error {
+
+	var reverseKeyFunc storage.ReverseKeyFunc
 	if e.KeyRootFunc == nil && e.KeyFunc == nil {
-		if isNamespaced {
```
Move refactor to separate PR

shadowofs replied:
Thanks for the suggestion!
I've split the refactor into a separate PR: #138699 
Will rebase this PR after that merges.

shadowofs commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -1654,7 +1700,7 @@ func (e *Store) CompleteWithOptions(options *generic.StoreOptions) error {
 		keyFuncs = defaultStoreKeyFuncs(prefix, isNamespaced)
 	} else {
-		keyFuncs = newStoreKeyFuncs(isNamespaced, e.KeyRootFunc, e.KeyFunc)
+		keyFuncs = newStoreKeyFuncs(isNamespaced, e.KeyRootFunc, e.KeyFunc, nil)
```
For custom `KeyRootFunc/KeyFunc`, the reverse key function remains nil because we cannot reliably infer the reverse mapping. That path is still safe: `WatchWithoutPrevKV` checks for a nil `reverseKeyFunc` and returns an internal error instead of using it.
https://github.com/kubernetes/kubernetes/pull/131862/changes#diff-6c00e4dbc36a123cb4ad0828c8aa43d026159563a7fbd1887a3445d6c341dd47R134-R136

shadowofs commented:
@serathius PTAL when you have time. All presubmits on the latest head are green, and I believe the review comments have been addressed.

serathius commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store_test.go:
Should we update this test too?

serathius commented:
I think the change looks good to me, however I don't have approval outside storage. Do you have any other recommendations how we can split the PR to:
* Make it more readable for approver without full context.
* Minimize changes outside of storage

Smaller more focused review will help move things forward.

shadowofs commented:
@serathius Thanks, that makes sense.

I split out the preparatory reverse key function plumbing into #139131. That PR only wires the reverse key function through registry/storage setup and adds the related tests; it does not change watch behavior.

Once #139131 merges, I will rebase this PR on top of it so this PR can stay focused on WatchWithoutPrevKV itself: feature gate, cacher -> etcd watch option, etcd3 watcher behavior, and the storage/cacher tests.

serathius commented:
Prepared a benchmark for watch cache https://github.com/kubernetes/kubernetes/pull/139440

Run it for this PR, the results:
* **Write Throughput Increase:** Disabling PrevKV watch option under active background watchlist load increases concurrent Delete/Create throughput by +34% (from ~25k to ~34k writes/second).
* **Overall Geomean Increase:** Overall write throughput across all scenarios increased by +18.21% (from 37.05k to 43.80k writes/second), aligning with the reported ~20% improvement in the PR.

```
                                                                                 │ staging/src/k8s.io/apiserver/pkg/storage/cacher/baseline.bench │ staging/src/k8s.io/apiserver/pkg/storage/cacher/optimized.bench │
                                                                                 │                             sec/op                             │                  sec/op                   vs base               │
StoreWriteThroughput/Traffic=DeleteCreate/Background=None/UseIndex=false-24                                                          175.6µ ± 49%                              159.4µ ±  44%        ~ (p=0.699 n=6)
StoreWriteThroughput/Traffic=DeleteCreate/Background=Watcher/UseIndex=false-24                                                       193.1µ ± 43%                              224.8µ ±   6%  +16.36% (p=0.015 n=6)
StoreWriteThroughput/Traffic=DeleteCreate/Background=Watcher/UseIndex=true-24                                                        170.0µ ± 48%                              187.4µ ±  19%        ~ (p=0.394 n=6)
StoreWriteThroughput/Traffic=DeleteCreate/Background=Lister/UseIndex=false-24                                                        191.1µ ± 45%                              142.5µ ±  43%        ~ (p=0.240 n=6)
StoreWriteThroughput/Traffic=DeleteCreate/Background=Lister/UseIndex=true-24                                                         137.6µ ± 30%                              144.7µ ±  59%        ~ (p=0.699 n=6)
StoreWriteThroughput/Traffic=DeleteCreate/Background=WatchList/UseIndex=false-24                                                     164.4µ ± 16%                              146.3µ ±  43%        ~ (p=0.132 n=6)
StoreWriteThroughput/Traffic=DeleteCreate/Background=WatchList/UseIndex=true-24                                                      150.4µ ± 35%                              102.5µ ± 115%        ~ (p=0.589 n=6)
StoreWriteThroughput/Traffic=Patch/Background=None/UseIndex=false-24                                                                 82.90µ ± 54%                              38.44µ ±  57%  -53.63% (p=0.009 n=6)
StoreWriteThroughput/Traffic=Patch/Background=Watcher/UseIndex=false-24                                                              67.74µ ± 20%                              68.03µ ±  12%        ~ (p=0.937 n=6)
StoreWriteThroughput/Traffic=Patch/Background=Watcher/UseIndex=true-24                                                               37.00µ ± 23%                              36.21µ ±   3%        ~ (p=0.132 n=6)
StoreWriteThroughput/Traffic=Patch/Background=Lister/UseIndex=false-24                                                               39.92µ ± 40%                              39.00µ ±  88%        ~ (p=0.937 n=6)
StoreWriteThroughput/Traffic=Patch/Background=Lister/UseIndex=true-24                                                                30.12µ ±  6%                              33.97µ ±  24%  +12.77% (p=0.002 n=6)
StoreWriteThroughput/Traffic=Patch/Background=WatchList/UseIndex=false-24                                                            32.36µ ± 36%                              37.00µ ±   5%        ~ (p=0.065 n=6)
StoreWriteThroughput/Traffic=Patch/Background=WatchList/UseIndex=true-24                                                             32.33µ ± 12%                              34.91µ ±  22%   +8.00% (p=0.041 n=6)
geomean                                                                                                                              84.71µ                                    78.49µ          -7.35%

                                                                                 │ staging/src/k8s.io/apiserver/pkg/storage/cacher/baseline.bench │ staging/src/k8s.io/apiserver/pkg/storage/cacher/optimized.bench │
                                                                                 │                         seconds-delay                          │              seconds-delay                vs base               │
StoreWriteThroughput/Traffic=DeleteCreate/Background=None/UseIndex=false-24                                                          1.507 ±  98%                               2.171 ±  65%        ~ (p=0.180 n=6)
StoreWriteThroughput/Traffic=DeleteCreate/Background=Watcher/UseIndex=false-24                                                      651.3m ±  96%                              952.2m ± 144%  +46.21% (p=0.009 n=6)
StoreWriteThroughput/Traffic=DeleteCreate/Background=Watcher/UseIndex=true-24                                                       861.5m ± 139%                             1024.3m ± 198%        ~ (p=0.310 n=6)
StoreWriteThroughput/Traffic=DeleteCreate/Background=Lister/UseIndex=false-24                                                       588.1m ± 181%                              855.8m ± 140%        ~ (p=0.394 n=6)
StoreWriteThroughput/Traffic=DeleteCreate/Background=Lister/UseIndex=true-24                                                        685.1m ± 104%                             1560.5m ±  98%        ~ (p=0.132 n=6)
StoreWriteThroughput/Traffic=DeleteCreate/Background=WatchList/UseIndex=false-24                                                    861.9m ±  19%                             1209.5m ±  80%        ~ (p=0.394 n=6)
StoreWriteThroughput/Traffic=DeleteCreate/Background=WatchList/UseIndex=true-24                                                     724.6m ±  92%                              964.0m ± 217%        ~ (p=0.699 n=6)
StoreWriteThroughput/Traffic=Patch/Background=None/UseIndex=false-24                                                                92.30m ± 112%                              32.70m ± 271%        ~ (p=0.699 n=6)
StoreWriteThroughput/Traffic=Patch/Background=Watcher/UseIndex=false-24                                                             29.05m ± 118%                              40.16m ± 186%        ~ (p=0.310 n=6)
StoreWriteThroughput/Traffic=Patch/Background=Watcher/UseIndex=true-24                                                              33.47m ± 243%                              38.57m ± 188%        ~ (p=0.937 n=6)
StoreWriteThroughput/Traffic=Patch/Background=Lister/UseIndex=false-24                                                              32.37m ± 170%                              55.39m ±  47%        ~ (p=0.240 n=6)
StoreWriteThroughput/Traffic=Patch/Background=Lister/UseIndex=true-24                                                               31.57m ± 215%                              36.84m ±  15%        ~ (p=0.461 n=6)
StoreWriteThroughput/Traffic=Patch/Background=WatchList/UseIndex=false-24                                                           31.20m ±  25%                              32.28m ±  11%        ~ (p=0.937 n=6)
StoreWriteThroughput/Traffic=Patch/Background=WatchList/UseIndex=true-24                                                            33.98m ± 203%                              34.18m ±  12%        ~ (p=0.818 n=6)
geomean                                                                                                                             172.5m                                     212.2m         +23.02%

                                                                                 │ staging/src/k8s.io/apiserver/pkg/storage/cacher/baseline.bench │ staging/src/k8s.io/apiserver/pkg/storage/cacher/optimized.bench │
                                                                                 │                            writes/s                            │                writes/s                  vs base                │
StoreWriteThroughput/Traffic=DeleteCreate/Background=None/UseIndex=false-24                                                          27.01k ± 17%                              35.99k ±  5%   +33.24% (p=0.002 n=6)
StoreWriteThroughput/Traffic=DeleteCreate/Background=Watcher/UseIndex=false-24                                                       18.05k ± 20%                              21.46k ± 36%         ~ (p=0.065 n=6)
StoreWriteThroughput/Traffic=DeleteCreate/Background=Watcher/UseIndex=true-24                                                        28.01k ± 18%                              34.45k ± 21%   +22.99% (p=0.041 n=6)
StoreWriteThroughput/Traffic=DeleteCreate/Background=Lister/UseIndex=false-24                                                        23.64k ± 29%                              33.70k ±  9%   +42.53% (p=0.002 n=6)
StoreWriteThroughput/Traffic=DeleteCreate/Background=Lister/UseIndex=true-24                                                         23.16k ± 14%                              34.16k ± 18%   +47.50% (p=0.002 n=6)
StoreWriteThroughput/Traffic=DeleteCreate/Background=WatchList/UseIndex=false-24                                                     27.36k ± 11%                              33.95k ±  4%   +24.10% (p=0.002 n=6)
StoreWriteThroughput/Traffic=DeleteCreate/Background=WatchList/UseIndex=true-24                                                      25.73k ± 17%                              34.05k ±  8%   +32.31% (p=0.002 n=6)
StoreWriteThroughput/Traffic=Patch/Background=None/UseIndex=false-24                                                                 24.96k ± 81%                              53.51k ± 30%  +114.40% (p=0.009 n=6)
StoreWriteThroughput/Traffic=Patch/Background=Watcher/UseIndex=false-24                                                              30.64k ± 17%                              30.49k ± 13%         ~ (p=0.818 n=6)
StoreWriteThroughput/Traffic=Patch/Background=Watcher/UseIndex=true-24                                                               56.63k ± 20%                              57.76k ±  7%         ~ (p=0.589 n=6)
StoreWriteThroughput/Traffic=Patch/Background=Lister/UseIndex=false-24                                                               59.46k ± 15%                              60.12k ± 17%         ~ (p=0.937 n=6)
StoreWriteThroughput/Traffic=Patch/Background=Lister/UseIndex=true-24                                                                69.03k ±  2%                              60.50k ± 19%   -12.36% (p=0.002 n=6)
StoreWriteThroughput/Traffic=Patch/Background=WatchList/UseIndex=false-24                                                            63.80k ± 26%                              55.55k ±  5%         ~ (p=0.065 n=6)
StoreWriteThroughput/Traffic=Patch/Background=WatchList/UseIndex=true-24                                                             63.98k ±  8%                              58.77k ± 19%    -8.15% (p=0.009 n=6)
geomean                                                                                                                              34.80k                                    41.13k         +18.18%

                                                                               │ staging/src/k8s.io/apiserver/pkg/storage/cacher/baseline.bench │ staging/src/k8s.io/apiserver/pkg/storage/cacher/optimized.bench │
                                                                               │                         watch-events/s                         │              watch-events/s               vs base               │
StoreWriteThroughput/Traffic=DeleteCreate/Background=Watcher/UseIndex=false-24                                                     375.3k ± 83%                               511.0k ± 50%        ~ (p=0.180 n=6)
StoreWriteThroughput/Traffic=DeleteCreate/Background=Watcher/UseIndex=true-24                                                       468.3 ± 85%                                707.5 ± 58%        ~ (p=0.485 n=6)
StoreWriteThroughput/Traffic=Patch/Background=Watcher/UseIndex=false-24                                                            1.028M ± 57%                               1.248M ± 19%        ~ (p=0.180 n=6)
StoreWriteThroughput/Traffic=Patch/Background=Watcher/UseIndex=true-24                                                              262.2 ± 28%                                255.6 ± 20%        ~ (p=0.589 n=6)
geomean                                                                                                                            14.75k                                     18.43k        +24.93%

                                                                                 │ staging/src/k8s.io/apiserver/pkg/storage/cacher/baseline.bench │ staging/src/k8s.io/apiserver/pkg/storage/cacher/optimized.bench │
                                                                                 │                          list-calls/s                          │               list-calls/s                vs base               │
StoreWriteThroughput/Traffic=DeleteCreate/Background=Lister/UseIndex=false-24                                                         7.457 ± 39%                                6.450 ± 46%        ~ (p=0.937 n=6)
StoreWriteThroughput/Traffic=DeleteCreate/Background=Lister/UseIndex=true-24                                                          21.93 ± 22%                                24.48 ± 76%        ~ (p=0.589 n=6)
StoreWriteThroughput/Traffic=DeleteCreate/Background=WatchList/UseIndex=false-24                                                      2.838 ± 12%                                2.671 ±  7%        ~ (p=0.065 n=6)
StoreWriteThroughput/Traffic=DeleteCreate/Background=WatchList/UseIndex=true-24                                                       5.675 ± 29%                                4.581 ± 52%        ~ (p=0.093 n=6)
StoreWriteThroughput/Traffic=Patch/Background=Lister/UseIndex=false-24                                                                6.052 ± 56%                                6.358 ± 81%        ~ (p=0.589 n=6)
StoreWriteThroughput/Traffic=Patch/Background=Lister/UseIndex=true-24                                                                 89.59 ±  4%                                83.62 ±  9%   -6.66% (p=0.002 n=6)
StoreWriteThroughput/Traffic=Patch/Background=WatchList/UseIndex=false-24                                                            974.5m ±  4%                               789.6m ±  8%  -18.96% (p=0.002 n=6)
StoreWriteThroughput/Traffic=Patch/Background=WatchList/UseIndex=true-24                                                              5.382 ± 11%                                4.834 ±  9%        ~ (p=0.065 n=6)
geomean                                                                                                                               7.233                                      6.671         -7.77%

                                                                                 │ staging/src/k8s.io/apiserver/pkg/storage/cacher/baseline.bench │ staging/src/k8s.io/apiserver/pkg/storage/cacher/optimized.bench │
                                                                                 │                          list-objs/s                           │               list-objs/s                 vs base               │
StoreWriteThroughput/Traffic=DeleteCreate/Background=Lister/UseIndex=false-24                                                       1117.0k ± 39%                               965.6k ± 47%        ~ (p=0.937 n=6)
StoreWriteThroughput/Traffic=DeleteCreate/Background=Lister/UseIndex=true-24                                                          744.4 ± 22%                                807.5 ± 76%        ~ (p=0.699 n=6)
StoreWriteThroughput/Traffic=DeleteCreate/Background=WatchList/UseIndex=false-24                                                     113.5k ± 53%                               124.6k ± 46%        ~ (p=0.937 n=6)
StoreWriteThroughput/Traffic=DeleteCreate/Background=WatchList/UseIndex=true-24                                                       205.7 ± 33%                                152.0 ± 52%        ~ (p=0.065 n=6)
StoreWriteThroughput/Traffic=Patch/Background=Lister/UseIndex=false-24                                                               907.7k ± 56%                               953.8k ± 81%        ~ (p=0.589 n=6)
StoreWriteThroughput/Traffic=Patch/Background=Lister/UseIndex=true-24                                                                3.046k ±  4%                               2.760k ±  9%   -9.41% (p=0.002 n=6)
StoreWriteThroughput/Traffic=Patch/Background=WatchList/UseIndex=false-24                                                            223.3k ± 15%                               184.7k ±  9%  -17.26% (p=0.009 n=6)
StoreWriteThroughput/Traffic=Patch/Background=WatchList/UseIndex=true-24                                                              190.1 ±  6%                                167.6 ±  7%  -11.79% (p=0.002 n=6)
geomean                                                                                                                              14.78k                                     13.65k         -7.66%


```
